### PR TITLE
[uwtools_integration] Integrate run_fcst

### PR DESCRIPTION
**Pull request by christinaholtNOAA**
## DESCRIPTION OF CHANGES: 

These changes allow the forecast to be configured and run by UW Tools.

## CONTRIBUTORS (optional): 

@NaureenBharwaniNOAA

### CODE REVIEW

File: parm/diag_table.FV3_GFS_v15_thompson_mynn_lam3km
Comment (christinaholtNOAA) [inline parm/diag_table.FV3_GFS_v15_thompson_mynn_lam3km:2]: Just making all these templates a bit more consistent with uwtools' driver lingo. The FV3 Driver will always pass its `cycle` datetime object. Users can pass any other variables via the `template_values` section in the fv3 YAML.

File: parm/fixed_files_mapping.yaml
Comment (christinaholtNOAA) [inline parm/fixed_files_mapping.yaml:6]: This file was a lot of shell-based stuff that didn't work with "flattening" the YAML in the run scripts. The relevant information has been moved to the FV3 YAML.

Two sections remain here. The first describes the list of variables that come out of make_sfc_climo, which the UW Driver could absorb once the `output` method is instantiated on that class. The second describes the set of required fix files the Thompson mp scheme needs. Perhaps those belong over in `ccpp_suites_defaults.yaml`? I'll take another peek at that.

File: parm/input.nml.FV3
Comment (christinaholtNOAA) [inline parm/input.nml.FV3:208]: UW Config does not like empty sections!

File: parm/model_configure
Comment (christinaholtNOAA) [inline parm/model_configure:1]: This file is now handled by YAML directly in config_defaults.yaml

File: parm/ufs.configure
Comment (christinaholtNOAA) [inline parm/ufs.configure:33]: Put this here for future work after discussing it as a possibility. I'll need it when I do the full test suite debugging/mods.

File: parm/wflow/coldstart.yaml
Comment (christinaholtNOAA) [inline parm/wflow/coldstart.yaml:121]: A little clean-up after the previous merge.

File: scripts/exregional_integration_test.py
Comment (christinaholtNOAA) [inline scripts/exregional_integration_test.py:118]: It's now YAML-compatible!

File: scripts/exregional_make_grid.sh
Comment (christinaholtNOAA) [inline scripts/exregional_make_grid.sh:116]: This will get to go away very soon!

File: scripts/exregional_make_grid.sh
Comment (christinaholtNOAA) [inline scripts/exregional_make_grid.sh:743]: I don't know why this was needed at all. 🤔

File: scripts/run_fcst.py
Comment (christinaholtNOAA) [inline scripts/run_fcst.py:103]: I'd like to follow up with this functionality in a PR of its own. I'm not sure that any of the tests actually test it, so it might need a decent bit of extra creativity to get it right.

File: ush/ccpp_suites_defaults.yaml
Comment (christinaholtNOAA) [inline ush/ccpp_suites_defaults.yaml:31]: All the namelist settings were moved over from parm/FV3.input.yaml.

File: ush/ccpp_suites_defaults.yaml
Comment (christinaholtNOAA) [inline ush/ccpp_suites_defaults.yaml:29]: These files are needed when the HRRR's gravity wave drag scheme is used.

File: ush/config_defaults.yaml
Comment (christinaholtNOAA) [inline ush/config_defaults.yaml:562]: A lot of these are just unnecessary re-directs to build up paths to files that are housed in SRW and where they end up in relation to the experiment.

File: ush/config_defaults.yaml
Comment (christinaholtNOAA) [inline ush/config_defaults.yaml:1766]: I've noticed that I need to use this `// 1` trick in cases where I need to ensure that the referenced variable has an actual numerical value before it gets substituted.

File: ush/config_defaults.yaml
Comment (christinaholtNOAA) [inline ush/config_defaults.yaml:1783]: This came over in a much more straightforward way from the `parm/fixed_files_mapping.yaml` file.

File: ush/config_defaults.yaml
Comment (christinaholtNOAA) [inline ush/config_defaults.yaml:1805]: We don't yet have a driver that handles this guy. For now, it will just continue to be handled as the template SRW already had and I followed UW-like structure for that.

File: ush/config_defaults.yaml
Comment (christinaholtNOAA) [inline ush/config_defaults.yaml:2090]: I moved this logic to the script because it was the first instance of needing a `!bool`. I should put it back and use our new flag.

File: ush/config_defaults.yaml
Comment (christinaholtNOAA) [inline ush/config_defaults.yaml:2395]: I mapped all these directly to the appropriate namelist settings in `stochastic_params.yaml`.

File: ush/config_defaults_aqm.yaml
Comment (christinaholtNOAA) [inline ush/config_defaults_aqm.yaml:14]: This section should only appear when AQM is turned on.

File: ush/generate_FV3LAM_wflow.py
Comment (christinaholtNOAA) [inline ush/generate_FV3LAM_wflow.py:26]: I blacked/linted this one.

File: ush/generate_FV3LAM_wflow.py
Comment (christinaholtNOAA) [inline ush/generate_FV3LAM_wflow.py:290]: I have no idea how this ever got into this script instead of living in `setup.py`. Anyway, it's now in the appropriate config files (or maybe a tiny bit in setup.py).

File: ush/generate_FV3LAM_wflow.py
Comment (christinaholtNOAA) [inline ush/generate_FV3LAM_wflow.py:505]: SRW puts a "base namelist" or several depending on the experiment in the experiment directory. I've kept that, but it's only the "base" as it relates to the CCPP suite. All other changes will be applied from the `update_values` section of the `config_defaults.yaml` file.

File: ush/link_fix.py
Comment (christinaholtNOAA) [inline ush/link_fix.py:416]: This file will likely go away with the make_grid integration.

File: ush/machine/hercules.yaml
Comment (christinaholtNOAA) [inline ush/machine/hercules.yaml]: Turns out the `|int` didn't help. Taking that out.

**Comment by christinaholtNOAA** [comment]
Switching to in-progress. I forgot to check the test files for relevant changes.

File: parm/fixed_files_mapping.yaml
Comment (maddenp-cu) [reply parm/fixed_files_mapping.yaml:6]: Are those Thompson mp input files eventually involved in a `uw fs copy` or `link` action? Mostly just curious how the list is used w.r.t. `uwtools`.

File: parm/input.nml.FV3
Comment (maddenp-cu) [reply parm/input.nml.FV3:208]: Interesting. Is this something that we might need to support? On the one hand it's kind of pointless to have an empty namelist; on the other hand I can see that someone might want one as a placeholder indicating that they are _aware_ of the namelist and are explicitly not overriding defaults in the Fortran code, or even just as documentation of what namelists exist in the code.

File: scripts/exregional_make_grid.sh
Comment (maddenp-cu) [inline scripts/exregional_make_grid.sh:750]: lol

File: scripts/run_fcst.py
Comment (maddenp-cu) [inline scripts/run_fcst.py:40]: Do we maybe have multiple copies of this in SRW now? I think I see at least one more in `chgres_cube.py`. Should we think about a `support.py` or `common.py` module to factor out repetition between the scripts?

File: scripts/run_fcst.py
Comment (maddenp-cu) [inline scripts/run_fcst.py:53]: Any reason not to use `uw fs link` or its API equivalent for this? Maybe one reason: This unlinks files first and re-does work vs the `uwtools` idempotent behavior. Should we think about adding an option to the `fs` actions to force them to remove and then recreate their targets?

File: scripts/run_fcst.py
Comment (maddenp-cu) [inline scripts/run_fcst.py:89]: In the `parse_args()` functions of the new scripts we're adding, we could also DRY thinks out in a `support` module, e.g. have
```
def add_argument_cycle(parser):
    parser.add_argument(
        "--cycle",
        ...
```
and then here just call `support.add_argument_cycle(parser)`. It seems like we're going to have a lot of repeated arguments across scripts where they really should all be the same. In some cases we might want
```
def add_argument_foo(parser, required: bool = True):
    parser.add_argument(
        ...
        required=required,
        ...
```
if there are arguments that are required for some scripts and not for others.

File: ush/config_defaults.yaml
Comment (maddenp-cu) [reply ush/config_defaults.yaml:1766]: What else might it be, and what does `// 1` trigger if it's the wrong thing -- exit with exception and stacktrace? Is this meant to avoid a problem with `!int`?

File: ush/setup.py
Comment (maddenp-cu) [inline ush/setup.py:575]: Is this really meant to be a multiline comment? I mean, should the comment really span multiple lines on the console? I'm guessing not... If there are multiple lines, I think they should probably be subject to separate `logging.info()` calls, as otherwise e.g. `grep`ing for a log message of interest may only get part of the message, if it's split over lines. IMO long log messages are fine. If it's just that the line is too long in Python, I think this could just be like
```
>>> logging.info(
...     "This is a very long log message but I want to control how long"
...     " it appears in Python because it is difficult to read very long"
...     " lines that wrap around"
... )
[2024-11-25T16:48:16]     INFO This is a very long log message but I want to control how long it appears in Python because it is difficult to read very long lines that wrap around
```
Any of the strings could be f-strings.

File: ush/setup.py
Comment (maddenp-cu) [inline ush/setup.py]: I'm not sure this captures the full intent, but could this just be
```
msg = f'''
Ensemble verification cannot be run unless running in ensemble mode:
    DO_ENSEMBLE = "{do_ensemble}"
Ensemble verification tasks:
    {task_str}
Please set DO_ENSEMBLE to True or remove ensemble vx tasks from the workflow.
'''
raise Exception(dedent(msg).strip())
```
?

**Review by maddenp-cu — APPROVED** [review]
I made a couple comments and asked a couple questions, and will follow-up on answers and/or do another pass if it's helpful, though I don't see any reason this couldn't be merged. Some of my suggestions might apply to past and future PRs on this branch, too, and might be better addressed in a separate PR once everything is ready -- for example, suggestions to DRY things out across scripts.

File: parm/fixed_files_mapping.yaml
Comment (christinaholtNOAA) [reply parm/fixed_files_mapping.yaml:6]: Yes. They do eventually get added to the `files_to_link` section of the fv3 UW YAML. Instead of listing the names of CCPP Suites that use Thompson MP, though, SRW calls a function that loads the XML to find out if Thompson is used. I could definitely clean up this section and move it over to CCPP suites to put it closer to its other relevant settings.

File: parm/input.nml.FV3
Comment (christinaholtNOAA) [reply parm/input.nml.FV3:208]: I'm totally neutral on this, maybe even leaning away from it. The weather model has so many sections that may never be loaded. A user needs to be pretty familiar with the model if they want to make any changes themselves.

File: scripts/exregional_make_grid.sh
Comment (christinaholtNOAA) [reply scripts/exregional_make_grid.sh:750]: This will never make it in this form to the main repo, but it felt good anyway.

File: scripts/run_fcst.py
Comment (christinaholtNOAA) [reply scripts/run_fcst.py:40]: Historically, we've had to put those types of files in the `ush` directory. If we did this, I'd want it in the same scripts directory so that we don't have to manage a `PYTHONPATH`.

Unless, of course, you are talking about opening it to the uwtools API. Then, yes, I think that would be great. I've essentially copied this directly from uwtools anyway.

I'd like to hear your thoughts on both options.

File: scripts/run_fcst.py
Comment (christinaholtNOAA) [reply scripts/run_fcst.py:53]: The unlinking was the main reason, I think. It could be nice to stay consistent. This solution, for example, doesn't do any logging on it's own, so I had to add it in. 

I wouldn't hate an "unlink" option for our `fs` tool.

File: scripts/run_fcst.py
Comment (christinaholtNOAA) [reply scripts/run_fcst.py:89]: I'm not entirely sure that the SRW devs would use that consistently. I can see a future where SRW would have some scripts that we contribute that would use that kind of tool, and others that use `argparse` directly. I doubt this would be an enforceable construct to the community. 

I'm stuck here in the middle of explicit (in the script) vs abstracting out the commonalities and which would be more friendly toward SRW developers.

File: scripts/run_fcst.py
Comment (maddenp-cu) [reply scripts/run_fcst.py:40]: I hadn't thought of exposing the `uwtools` version. Hmm... We could consider something like `uwtools.api.support` or `uwtools.api.utils` where we could expose otherwise-internal functions that we determine are more widely useful. Of course, that increases the API surface area and the risk of breaking changes when we need to change previously-internal functionality to suite still-internal needs. I think maybe we should keep this in mind but maybe wait until we have some kind of critical mass -- like 3 or more functions `uwtools` functions that we find ourselves reimplementing in apps -- before going this route.

I was thinking of just something alongside the other scripts that they could `import`.

The common module should not have executable bits set... :)

File: scripts/run_fcst.py
Comment (maddenp-cu) [reply scripts/run_fcst.py:53]: This might be related to [this](https://github.com/orgs/ufs-community/projects/56/views/10?pane=issue&itemId=76720127&issue=ufs-community%7Cuwtools%7C605), which would require reliable `output()` methods to know what to remove, but this is simpler, as there's only a file or symlink to `rm` -- or, maybe, a `rm -rf` for a directory -- for `fs` targets. Implementing `--reset` for `fs` might be a good start.

On the other hand, I do wonder about mixing behaviors: In some parts of an app's automation, logic (like this) that repeatedly clobbers existing on-disk state; and in other parts (like invocation of drivers) logic that is idempotent. Maybe that's ok, but I wonder if users will understand which part has which behavior.

File: scripts/run_fcst.py
Comment (maddenp-cu) [reply scripts/run_fcst.py:89]: I'd imagine that developers would mostly copy existing patterns. If the pattern they observe is copy-paste, we'll certainly get more of that -- and then subtle drift over time as ad hoc changes are made without regard to the larger design. If they see much shorter code that uses something `import`ed, I suppose they'll copy that. I feel like we can at least try to set a good example and establish a useful pattern. Whether or not what I'm suggesting is "good" or "useful" -- and I agree that there's a balance between explicit and DRY -- is your call.

> I doubt this would be an enforceable construct to the community.

That's what PRs are for, but if reviewers won't request adherence to convention, there's not much we can do.

File: ush/config_defaults.yaml
Comment (christinaholtNOAA) [reply ush/config_defaults.yaml:1766]: Sorry I missed this comment earlier.

It's meant to avoid a problem when there are multiple steps to render values. In a small example:

```
a: 2
b: 7
foo: !int "{{ a * b }}"
bar: !int "{{ foo }}"
``` 

Realizing this config:

```
$ uw config realize -i example.yaml --output-format yaml
a: 2
b: 7
foo: 14
bar: !int '!int 14'
```

Adding the `// 1` to `bar: !int "{{ foo // 1 }}"`  gives the intended `bar: 14` result. I believe this allows the `bar` value to hold off until `foo` has an actual integer value. The integer divide by 1 raises an exception except for when the numerator is actually a number, and our renderer allows for broad exceptions that result in the expression _not_ being rendered. 

It's a cheeky workaround to have the values wait until it's appropriate to render the expression.

File: scripts/run_fcst.py
Comment (christinaholtNOAA) [reply scripts/run_fcst.py:53]: This came in handy during the debugging step when we were trying to get the linking correct. We could add back the idempotence here by just saying if it exists, don't try to re-link. It is external to the rundir though, so this behavior could send users on a wild goose chase through Timbuktu to unlink everything a given SRW script might try to link.

File: scripts/run_fcst.py
Comment (christinaholtNOAA) [reply scripts/run_fcst.py:89]: > but if reviewers won't request adherence to convention, there's not much we can do.

That's been my main concern.

We get such a mix of developers in the apps that some will go for copy/paste within SRW, and others will try to write something from scratch (that's sometimes quite thrilling) or copy/paste it from somewhere else to try to slot it in. That doesn't work yet (give us time)! 

We could definitely try it out as a potential clean-up item.

File: ush/setup.py
Comment (christinaholtNOAA) [reply ush/setup.py:575]: I really like this argument. I think it will be out of scope for the set up main SRW integration PRs, but I'd happily come back and take out this nonsense in a cleanup  PR!

Some SRW developers really want to control how messages show up in the console, too.

File: ush/setup.py
Comment (christinaholtNOAA) [reply ush/setup.py]: I didn't even notice this one. Absolutely.

File: ush/config_defaults.yaml
Comment (maddenp-cu) [reply ush/config_defaults.yaml:1766]: IMO it would be better if we could deal with the core issue in `uwtools` instead of using this workaround, whose meaning may not be clear to users/maintainers. It's pretty subtle. And there are now a lot of instances of it... Don't wait for me, but let me see if there's a way we could address this correctly. Thanks for the minimal reproducer!

File: scripts/run_fcst.py
Comment (maddenp-cu) [reply scripts/run_fcst.py:53]: > We could add back the idempotence here by just saying if it exists, don't try to re-link.

Would that then be equivalent to using `uw fs link` (CLI or API)? I'm just angling for using `uwtools` functionality where possible, of course. I'll leave it at that -- perfect is not the enemy of good! (And maybe this is even better, I don't know.)

**Pull request merged**
